rllib/utils/metrics/metrics_logger.py

```diff
@@ -238,20 +238,39 @@ def reduce(
             objects if `return_stats_obj=True` or primitive values, carrying no
             reduction and history information, if `return_stats_obj=False`.
         """
+        # For better error message, catch the last key-path (reducing of which might
+        # throw an error).
+        PATH = None
+
+        def _reduce(path, stats):
+            nonlocal PATH
+            PATH = path
+            return stats.reduce()
+
         # Create a shallow copy of `self.stats` in case we need to reset some of our
         # stats due to this `reduce()` call (and the Stat having self.clear_on_reduce
         # set to True). In case we clear the Stats upon `reduce`, we get returned a
         # new empty `Stats` object from `stat.reduce()` with the same settings as
         # existing one and can now re-assign it to `self.stats[key]` (while we return
         # from this method the properly reduced, but not cleared/emptied new `Stats`).
-        if key is not None:
-            stats_to_return = self._get_key(key).copy()
-            self._set_key(
-                key, tree.map_structure(lambda s: s.reduce(), stats_to_return)
+        try:
+            if key is not None:
+                stats_to_return = self._get_key(key).copy()
+                self._set_key(
+                    key, tree.map_structure_with_path(_reduce, stats_to_return)
+                )
+            else:
+                stats_to_return = self.stats.copy()
+                self.stats = tree.map_structure_with_path(_reduce, stats_to_return)
+        # Provide proper error message if reduction fails due to bad data.
+        except Exception as e:
+            raise ValueError(
```

Comment:
Here I wonder what happens when a user resets in between two training steps a key.

Reply:
Great question. Users should - ideally - never use this API and let RLlib determine, when to call `.reduce()` on a MetricsLogger object. They should only call `.peek()` to see individual reduced stats at the moment w/o actually altering/reducing the underlying stats object.

Reply:
Maybe I need to make this more clear in the docstring ...

Reply:
done